src/rust/ensogl/lib/core/src/display/shape/primitive/system.rs

```diff
@@ -117,44 +130,192 @@ impl display::Object for ShapeSystem {
+    /// New shape instantiation. Used to bind a shape to a specific scene implementation.
+    fn instantiate(&self, shape:&Self::DynamicShape) -> attribute::InstanceIndex;
+}
+
+/// Abstraction for every entity which is associated with a shape system (user generated one). For
+/// example, all defined shapes are associated with a shape system, and thus they implement this
+/// trait.
+pub trait KnownShapeSystemId {
+    /// The ID of a user defined shape system.
+    fn shape_system_id() -> ShapeSystemId;
+}
+
 
 
 // =============
 // === Shape ===
 // =============
 
-/// Type for every shape with automatic attribute management. The easiest way to define such a
+/// Type for every shape bound to a specific scene and GPU buffers. The easiest way to define such a
 /// shape is by using the `define_shape_system` macro.
 pub trait Shape : display::Object + CloneRef + Debug + Sized {
     /// The shape system instance this shape belongs to.
-    type System : ShapeSystemInstance<Shape=Self>;
-    /// Accessor for the underlying sprites.
-    fn sprites(&self) -> Vec<&Sprite>;
+    type System : StaticShapeSystemInstance<Shape=Self>;
+    /// Accessor for the underlying sprite.
+    fn sprite(&self) -> &Sprite;
+}
+
+/// Type for every shape which can, but does not have to be bound to a specific scene and GPU
+/// buffers. Dynamic shapes can be created freely and will be bound to a scene after being attached
+/// as scene children and an update frame event will be emitted.
+///
+/// Dynamic shapes contain copy of all shape parameters and use them to set up the GPU parameters
+/// on bound.
+///
+/// The easiest way to define such a shape is by using the `define_shape_system` macro.
+pub trait DynamicShape : display::Object + CloneRef + Debug + Sized {
+    /// The static version of the shape. Dynamic shapes can be associated with one or more static
+    /// shapes after they are placed on the stage and initialized.
+    type StaticShape : Shape;
+    /// The shape system instance this shape belongs to.
+    type System : DynShapeSystemInstance<DynamicShape=Self>;
+    /// Constructor.
+    fn new(logger:impl AnyLogger) -> Self;
+    /// Accessor for the underlying sprite, if the shape is initialized.
+    fn sprites(&self) -> Vec<Sprite>;
+    /// The "canvas" size of the shape. It defines the bounding-box for the shape drawing area.
+    fn size(&self) -> &DynamicParam<sprite::Size>;
+    /// Check if given pointer-event-target means this object.
+    fn is_this_target(&self, target:display::scene::PointerTarget) -> bool {
+        self.sprites().into_iter().any(|sprite| sprite.is_this_target(target))
+    }
 }
 
+/// Internal utilities for managing [`DynamicShape`]s. You would normally not need to ever use it
+// explicitly, however, it is exposed as public interface as it is required for user-defined shape
+// systems. Again, instead of implementing shape systems from scratch, you'd rather use the
+// `define_shape_system!` macro.
```

Comment:
```suggestion
/// Internal utilities for managing [`DynamicShape`]s. You would normally not need to ever use it
/// explicitly, however, it is exposed as public interface as it is required for user-defined shape
/// systems. Again, instead of implementing shape systems from scratch, you'd rather use the
/// `define_shape_system!` macro.
```
And don't use "again" as we can't rely on any order the user will read the docs.